keras_nlp/utils/text_generation.py

```diff
@@ -288,63 +298,91 @@ def token_probability_fn(inputs):
     ```
 
     """
-    if not tf.executing_eagerly():
-        raise RuntimeError(
-            "`keras_nlp.utils.beam_search` currently requires an eager "
-            "execution context. Please call `beam_search` outside "
-            "tf.function or run `tf.config.run_functions_eagerly(True)` to run "
-            "tf.function in eager mode."
-        )
     if num_beams <= 0:
         raise ValueError(
             f"`num_beams` should be strictly positive. Received: `num_beams={num_beams}`."
         )
-
     prompt = _validate_prompt(prompt)
 
-    if prompt.shape[-1] == 0:
-        raise ValueError(
-            "Length of `prompt` is 0, please provide a non-empty `prompt`."
-        )
-
     input_is_1d = prompt.shape.rank == 1
     if input_is_1d:
         prompt = prompt[tf.newaxis, :]
-    _validate_token_probability_fn(token_probability_fn, prompt)
 
-    batch_size, length = prompt.shape
+    batch_size, length = _get_prompt_shape(prompt)
+    prompt, mask = _pad_prompt(prompt, max_length)
+
+    vocab_size = _validate_token_probability_fn(
+        token_probability_fn, prompt, return_vocab_size=True
+    )
+
     if length >= max_length:
         return tf.squeeze(prompt) if input_is_1d else prompt
+    # Initialize beam with shape `(batch_size, num_beams, length)`.
+    beams = tf.repeat(tf.expand_dims(prompt, axis=1), num_beams, axis=1)
+
+    # Initialize beams_prob with shape `(batch_size, num_beams)`
+    beams_prob = tf.zeros([batch_size, 1], dtype=tf.float32)
+    if num_beams > 1:
+        beams_prob = tf.concat(
+            [beams_prob, tf.fill((batch_size, num_beams - 1), -3.40282e38)],
+            axis=-1,
+        )
+
+    def one_step(beams, beams_prob, length):
+        truncated_beams = beams[..., :length]
+
+        # logits is of shape [batch_size, num_beams * vocab_size]
+        logits = tf.reshape(
+            token_probability_fn(
```

Comment:
Are we changing the expectation around the `token_probability_fn` from what it was? it should support a input with a first dim of batch_size * num_beams of something like that?

That's probably ok, especially if we have performance data to back this up, but we should make sure to clearly document expectation for the function.

Reply:
Ideally, that shouldn't matter, right? Isn't `token_probability_fn` a function which just takes a number of samples and returns the set of next tokens for every sample? Won't most users just do a forward pass on the model, in which case, the number of samples should not matter? 

What is the alternative here? `tf.map_fn` over the beams? One potential issue with the current reshaping solution could be memory issues, which won't happen if we iterate over the beams. The downside is of course, worse performance (probably, since we are using a `map_fn`).

Reply:
I think it probably makes sense to use `token_probability_fn` as you do here, this is probably just a documentation problem.

We are changing this function (which is user inputted), to have a different shape expectation than all the other utilities fns. A good model does not need to encode a fixed batch size, but that won't stop some users from building models that barf on a different batch size.. We should make sure to document our shape expectations.

Reply:
Great! Will mention it in the documentation